File: src/main/java/com/github/tianchenghang/ai/tools/FileDeleteTool.java
package com.github.tianchenghang.ai.tools;

import cn.hutool.json.JSONObject;
import com.github.tianchenghang.constants.AppConstant;
import dev.langchain4j.agent.tool.P;
import dev.langchain4j.agent.tool.Tool;
import dev.langchain4j.agent.tool.ToolMemoryId;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Slf4j
@Component
public class FileDeleteTool extends BaseTool {
  @Tool("Delete file at the specified path")
  public String deleteFile(@P("Relative file path") String relativeFilepath, @ToolMemoryId Long appId) {
    try {
      var path = Paths.get(relativeFilepath);
      if (!path.isAbsolute()) {
        var projectDirname = "vite_project_" + appId;
        var projectRoot = Paths.get(AppConstant.CODE_OUTPUT_ROOT_DIR, projectDirname);
        path = projectRoot.resolve(relativeFilepath);
      }

      if (!Files.exists(path) || !Files.isRegularFile(path)) {
        return "File not found or not a file, cannot delete: " + relativeFilepath;
      }
      var filename = path.getFileName().toString();
      if (isImportantFile(filename)) {
        return "File deletion not allowed: " + relativeFilepath;
      }
      Files.delete(path);
      log.info("File deleted successfully: {}", path.toAbsolutePath());
      return "File deleted successfully: " + relativeFilepath;
    } catch (IOException e) {
      var errorMessage = String.format("File deletion failed: %s, error: %s", relativeFilepath, e.getMessage());
      log.error(errorMessage, e);
      return errorMessage;
    }
  }

  private boolean isImportantFile(String filename) {
    var importantFiles =
        new String[] {
          "package.json",
          "package-lock.json",
          "yarn.lock",
          "pnpm-lock.yaml",
          "vite.config.js",
          "vite.config.ts",
          "tsconfig.json",
          "tsconfig.app.json",
          "tsconfig.node.json",
          "index.html",
          "main.js",
          "main.ts",
          "App.tsx",
          "App.vue",
          ".gitignore",
          "README.md"
        };
    for (var file : importantFiles) {
      if (file.equalsIgnoreCase(filename)) {
        return true;
      }
    }
    return false;
  }

  @Override
  public String getToolName() {
    return "FileDelete";
  }

  @Override
  public String getDisplayName() {
    return "File Delete";
  }

  @Override
  public String generateToolExecuteResult(JSONObject arguments) {
    var relativeFilepath = arguments.getStr("relativeFilepath");
    return String.format("Invoke tool: %s %s", getDisplayName(), relativeFilepath);
  }
}

File: server/src/ai/tools/file-delete.tool.ts
import * as fs from 'fs';
import * as path from 'path';
import { BaseTool, ToolResult } from './base-tool';
import { tool } from '@langchain/core/tools';
import { z } from 'zod';

export class FileDeleteTool extends BaseTool {
  name = 'FileDelete';
  description =
    'Delete a file at the specified path. Some critical files are protected.';

  private readonly protectedFiles = [
    'package.json',
    'vite.config.ts',
    'vite.config.js',
    'index.html',
    'tsconfig.json',
    'tsconfig.node.json',
  ];

  async execute(params: { filePath: string }): Promise<ToolResult> {
    try {
      const fullPath = path.resolve(this.workDir, params.filePath);
      if (!fullPath.startsWith(path.resolve(this.workDir))) {
        return { success: false, message: 'Path traversal not allowed' };
      }

      const fileName = path.basename(params.filePath);
      if (this.protectedFiles.includes(fileName)) {
        return {
          success: false,
          message: `Cannot delete protected file: ${fileName}`,
        };
      }

      if (!fs.existsSync(fullPath)) {
        return {
          success: false,
          message: `File not found: ${params.filePath}`,
        };
      }

      fs.unlinkSync(fullPath);
      return { success: true, message: `File deleted: ${params.filePath}` };
    } catch (error: any) {
      return {
        success: false,
        message: `Failed to delete file: ${error.message}`,
      };
    }
  }

  toLangChainTool() {
    return tool(
      async (input: { filePath: string }) => {
        const result = await this.execute(input);
        return result.message;
      },
      {
        name: this.name,
        description: this.description,
        schema: z.object({
          filePath: z.string().describe('Relative file path to delete'),
        }),
      },
    );
  }
}
